### ISSUE TITLE
Parameterize node traversal helper methods over desired return type

### ISSUE THREAD

**Issue by jdm** [servo/servo#15270]
There's a proof-of-concept of this work in #14451, but the author does not have time to finish it. It's a worthwhile improvement, so someone else taking it over would be great!

**Comment by DominoTree** [comment]
I'll take this one!

**Comment by KiChjang** [comment]
@DominoTree Okay, but just to be sure, do you know what we're investigating in this issue?

**Comment by DominoTree** [comment]
@KiChjang I think so - from the proof-of-concept the basic idea seems pretty simple - adding a type parameter to generic iterators moving up and down and sideways through the DOM to return specified types - but I have a feeling implementation will be more involved

Does that sound like I'm on the right track?

**Comment by KiChjang** [comment]
Yeah, please don't hesitate to ask us any questions you might have during implementation here or in #servo at `irc.mozilla.org`!

**Comment by DominoTree** [comment]
Hacked on this quite a bit based on the work in #14451 and I think I can have this wrapped up pretty shortly.  One quick question - 

We have the following iterators defined which we can specify types for:
- SiblingIterator (Previously NodeSiblingIterator)
- ReverseSiblingIterator (Previously NodeReverseSiblingIterator)
- LastChildIterator
- AncestorIterator
- TreeIterator
- FollowingNodeIterator
- PrecedingNodeIterator

Should we skip `FollowingNodeIterator` and `PrecedingNodeIterator`, or should they be renamed to something without `Node` in the name?

**Comment by jdm** [comment]
`FollowingIterator` and `PrecedingIterator` sound sensible to me.

**Comment by tamird** [comment]
011ff28aee34f59eb1e788de2666062c7b6dbff4 merged:

```
NodeSiblingIterator
ReverseSiblingIterator
LastChildIterator
AncestorIterator
```
into
```
SimpleNodeIterator
```

which leaves:
```
TreeIterator
SimpleNodeIterator
FollowingNodeIterator
PrecedingNodeIterator
```
Also, looking at #15503, it seems that there's some difficulty in seeding the new parameterized iterators because the root type may not be the same as the item type. Perhaps the current pattern of `traversal_method.filter_map(Root::downcast)` is good enough, or perhaps better and clearer?

**Comment by DominoTree** [comment]
Hmm, I may take another fresh whack at this

**Comment by KiChjang** [comment]
I'm not sure if this is still desired. Can anybody weigh in on this?

**Comment by DominoTree** [comment]
Sounds like it's more effort to use a parameterized version than to leave it as is

**Comment by KiChjang** [comment]
Removing the less easy label until we can figure out the best way to implement this (or not at all).
